sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,861 @@
+                Optional.ofNullable(watermarkEstimator.currentWatermark()));
+      }
+      try (Consumer<byte[], byte[]> consumer = consumerFactoryFn.apply(updatedConsumerConfig)) {
+        consumerSpEL.evaluateAssign(
+            consumer, ImmutableList.of(kafkaSourceDescription.getTopicPartition()));
+        long startOffset = tracker.currentRestriction().getFrom();
+        long expectedOffset = startOffset;
+        consumer.seek(kafkaSourceDescription.getTopicPartition(), startOffset);
+        ConsumerRecords<byte[], byte[]> rawRecords = ConsumerRecords.empty();
+
+        try {
+          while (true) {
+            rawRecords = consumer.poll(KAFKA_POLL_TIMEOUT.getMillis());
+            // When there is no records from the current TopicPartition temporarily, self-checkpoint
+            // and move to process the next element.
+            if (rawRecords.isEmpty()) {
+              return ProcessContinuation.resume();
+            }
+            for (ConsumerRecord<byte[], byte[]> rawRecord : rawRecords) {
+              if (!tracker.tryClaim(rawRecord.offset())) {
+                return ProcessContinuation.stop();
+              }
+              KafkaRecord<K, V> kafkaRecord =
+                  new KafkaRecord<>(
+                      rawRecord.topic(),
+                      rawRecord.partition(),
+                      rawRecord.offset(),
+                      consumerSpEL.getRecordTimestamp(rawRecord),
+                      consumerSpEL.getRecordTimestampType(rawRecord),
+                      ConsumerSpEL.hasHeaders() ? rawRecord.headers() : null,
+                      keyDeserializerInstance.deserialize(rawRecord.topic(), rawRecord.key()),
+                      valueDeserializerInstance.deserialize(rawRecord.topic(), rawRecord.value()));
+              int recordSize =
+                  (rawRecord.key() == null ? 0 : rawRecord.key().length)
+                      + (rawRecord.value() == null ? 0 : rawRecord.value().length);
+              avgOffsetGap
+                  .computeIfAbsent(kafkaSourceDescription.getTopicPartition(), k -> new MovingAvg())
+                  .update(expectedOffset - rawRecord.offset());
+              avgRecordSize
+                  .computeIfAbsent(kafkaSourceDescription.getTopicPartition(), k -> new MovingAvg())
+                  .update(recordSize);
+              expectedOffset = rawRecord.offset() + 1;
+              Instant outputTimestamp;
+              // The outputTimestamp and watermark will be computed by timestampPolicy, where the
+              // WatermarkEstimator should be a Manual one.
+              if (timestampPolicy != null) {
+                checkState(watermarkEstimator instanceof ManualWatermarkEstimator);
+                TimestampPolicyContext context =
+                    new TimestampPolicyContext(
+                        (long) ((HasProgress) tracker).getProgress().getWorkRemaining(),
+                        Instant.now());
```

Comment:
The backlog check time doesn't accurately reflect when the value was memoized. Not sure if this will cause an issue.

Reply:
If the timeout of the supplier is 5s, it wouldn't have a significant impact.